### DISCUSSION TITLE
Countdown timer for Cart

### DISCUSSION THREAD

**Discussion by Web10-Joris** [duncanmcclean/simple-commerce#1079] [Help]
What is the optimal method for implementing a countdown timer for a cart page? The cart is stored in the session, right? Can the creation timestamp of the cart/session be retrieved on the frontend to display a countdown timer? Then the `SESSION_LIFETIME` is set to 60 minutes. Once the timer expires, the page should refresh, resulting in an empty cart or a new session.

**Comment by duncanmcclean** [comment]
Out of curiosity, what's your use case for this? Why do you need a user's cart to be wiped out after a set amount of time?

By default, Simple Commerce's "cart" is stored in a cookie, rather than in the user's session. It looks like it might be possible for us to set an expiry date/time for the cookie but I wonder if it might be easier for you to just remove the line items from the customer's existing cart instead of creating a new cart completely, otherwise you might end up with lots of unused carts.

**Reply by Web10-Joris** [reply]
Indeed, I noticed that the default driver is cookies rather than sessions. We want to have a countdown timer in the cart/checkout. Perhaps it would be best to either remove the stored cookie from the browser after a certain time or set an expiry in the CookieDriver makeCart function. Or am I thinking too complicated for this use case? 

Additionally, I have another question related to this. Is the stock reserved when a user adds a specific item to the cart? If not, is this something I can implement? The countdown timer's purpose is to show how long a product is reserved for the customer, as we have numerous products with limited stock.

**Reply by duncanmcclean** [reply]
> Indeed, I noticed that the default driver is cookies rather than sessions. We want to have a countdown timer in the cart/checkout. Perhaps it would be best to either remove the stored cookie from the browser after a certain time or set an expiry in the CookieDriver makeCart function. Or am I thinking too complicated for this use case?

I'd accept a PR to explicitly set the expiry date & time of the cookie, if that's something you can do.

Alternatively, you could maybe delete the cart/remove the line items via JavaScript by hitting one of Simple Commerce's form endpoints with AJAX when the time is up.

* To delete the cart: `DELETE /!/simple-commerce/cart`
* To remove a line item: `DELETE /!/simple-commerce/cart-items/{itemId}`

> Additionally, I have another question related to this. Is the stock reserved when a user adds a specific item to the cart? If not, is this something I can implement? The countdown timer's purpose is to show how long a product is reserved for the customer, as we have numerous products with limited stock.

No, the stock counter is only deducted when the customer checks out. If you need anything on top of this, you'd need to build it custom.

**Reply by Web10-Joris** [reply]
> I'd accept a PR to explicitly set the expiry date & time of the cookie, if that's something you can do.

This would only involve something like this right? Will make a PR as soon as I have time for it. 
  ```
$cookieExpiration = Config::get('simple-commerce.cart.expiration', 0); // In minutes
Cookie::queue($this->getKey(), $cart->id, $cookieExpiration);
```

> No, the stock counter is only deducted when the customer checks out. If you need anything on top of this, you'd need to build it custom.

I think having an optional feature to enable this would be great. What would be your recommendation? 
Should I (try to) integrate this by adding metadata to each line-item and managing the stock reservation from there, or at a higher order level  (based on order created date + expiration)? 

Just some quick ideas: 
1. When a product X is added, a timestamp with the expiry date is also added to that line item.
2. The stock for product X is deducted.
3. A command checks at a set interval for products that are still in a 'Cart' order and have their expiry due.
4. If the item is expired, the stock for product X is restocked.

Or

1. An order is created and stock is directly deducted upon adding to the cart.
2. A command checks 'Cart' orders along with their Line Items. If the created_date + expiry is due, it restocks the line items (and remove them from the order).

In both scenarios, the stock isn't available at checkout, so that should also be handled differently (since it is now validating the availability on checkout). I wouldn't mind spending time on this (possibly for a future release), but would need some guidance on implementation.

**Reply by duncanmcclean** [reply]
> This would only involve something like this right? Will make a PR as soon as I have time for it.

Yes, I believe so.

> When a product X is added, a timestamp with the expiry date is also added to that line item.
> The stock for product X is deducted.
> A command checks at a set interval for products that are still in a 'Cart' order and have their expiry due.
> If the item is expired, the stock for product X is restocked.

This sounds like a good idea to me. 

I'm currently reconsidering how parts of Simple Commerce work for future versions so I'll have a think about potentially doing something similar in the future.
